tlv/stream.go

```diff
@@ -0,0 +1,259 @@
+package tlv
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+)
+
+var (
+	// ErrStreamNotCanonical signals that a decoded stream does not contain
+	// records sorting by monotonically-increasing type.
+	ErrStreamNotCanonical = errors.New("tlv stream is not canonical")
+
+	// ErrNoEncoder signals that a record does not have an encoding codec.
+	ErrNoEncoder = errors.New("no encoding for record")
+
+	// ErrNoDecoder signals that a record does not have a decoding codec.
+	ErrNoDecoder = errors.New("no decoding for record")
+)
+
+// ErrUnknownRequiredType is an error returned when decoding an unknown and even
+// type from a Stream.
+type ErrUnknownRequiredType Type
+
+// Error returns a human-readable description of unknown required type.
+func (t ErrUnknownRequiredType) Error() string {
+	return fmt.Sprintf("unknown required type: %d", t)
+}
+
+// RecordEncoder specifies a codec for serializing a given record.
+type RecordEncoder func(io.Writer, interface{}) error
+
+// RecordDecoder specifies a codec for deserializing given record. The final
+// argument is the expected length, and the method should return an error if
+// thee number of bytes read does not match this value.
+type RecordDecoder func(io.Reader, interface{}, uint64) error
+
+// Stream defines a TLV stream that can be used for encoding or decoding a set
+// of TLV Records.
+type Stream struct {
+	records []Record
+	buf     [8]byte
+}
+
+// NewStream creates a new TLV Stream given an encoding codec, a decoding codec,
+// and a set of known records.
+func NewStream(records ...Record) *Stream {
+	// Assert that the ordering of the Records is canonical and appear in
+	// ascending order of type.
+	var min Type
+	var overflow bool
+	for _, record := range records {
+		if overflow || record.typ < min {
+			panic(ErrStreamNotCanonical.Error())
+		}
+		if record.encoder == nil {
+			panic(ErrNoEncoder.Error())
+		}
+		if record.decoder == nil {
+			panic(ErrNoDecoder.Error())
+		}
+		if record.typ == math.MaxUint64 {
+			overflow = true
```

Comment:
If the final type is what overflows, then it appears we won't panic at all?

Reply:
if the final type is `math.MaxUint64` that's okay, so long as there isn't a number after it. `overflow` in this case signifies that a record hit `math.MaxUint64`, and any other records should be forbidden